calico_containers/docker_plugin.py

```diff
@@ -253,6 +237,75 @@ def unassign_ip(ip):
     return False
 
 
+def ipv4_and_gateway(ep):
+    # Get the gateway before trying to assign an address. This will avoid
+    # needing to backout the assignment if fetching the gateway fails.
+    try:
+        next_hop = client.get_default_next_hops(hostname)[4]
+    except KeyError as e:
+        app.logger.exception(e)
+        abort(500)
+
+    ip = assign_ip("v4")
+    app.logger.info("Assigned IPv4 %s", ip)
+
+    if not ip:
+        app.logger.error("Failed to allocate IPv4 for endpoint %s",
+                         ep.ep_id)
+        abort(500)
+
+    ip = IPNetwork(ip)
+    ep.ipv4_nets.add(ip)
+    ep.ipv4_gateway = next_hop
+
+
+def ipv6_and_gateway(ep):
+    try:
+        next_hop6 = client.get_default_next_hops(hostname)[6]
+    except KeyError:
+        app.logger.info("Couldn't find IPv6 gateway for endpoint %s. "
+                        "Skipping IPv6 assignment.",
+                        ep.ep_id)
+    else:
+        ip6 = assign_ip("v6")
+        if ip6:
+            ip6 = IPNetwork(ip6)
+            ep.ipv6_gateway = next_hop6
+            ep.ipv6_nets.add(ip6)
+        else:
+            app.logger.info("Failed to allocate IPv6 address for endpoint %s",
+                            ep.ep_id)
+
+
+def backout_ip_assignments(ep):
+    for ip in ep.ipv4_nets.union(ep.ipv6_nets):
+        # The unassignment is best effort. Just log if it fails.
+        if not unassign_ip(ip):
```

Comment:
This line is bugged.  `ipv4_nets` and `ipv6_nets` are IPNetworks, not IP addresses.  See `delete_endpoint` on the tip of `libnetwork-rebased` for the fix.

Also, `delete_endpoint` should use this method.